src/osd/PrimaryLogPG.cc

```diff
@@ -7094,14 +7148,58 @@ void PrimaryLogPG::write_update_size_and_usage(object_stat_sum_t& delta_stats, o
   modified.union_of(ch);
   if (write_full || offset + length > oi.size) {
     uint64_t new_size = offset + length;
-    delta_stats.num_bytes -= oi.size;
-    delta_stats.num_bytes += new_size;
+    if (!oi.has_extents()) {
+      delta_stats.num_bytes -= oi.size;
+      delta_stats.num_bytes += new_size;
+    }
     oi.size = new_size;
   }
+  if (length && oi.has_extents()) {
+     // count newly write bytes, exclude overwrites
+     interval_set<uint64_t> ne;
+     ne.insert(offset, length);
+     interval_set<uint64_t> overlap;
+     overlap.intersection_of(ne, oi.extents);
+     ne.subtract(overlap);
+     oi.extents.union_of(ne);
+     delta_stats.num_bytes += ne.size();
+  }
   delta_stats.num_wr++;
   delta_stats.num_wr_kb += SHIFT_ROUND_UP(length, 10);
 }
 
+void PrimaryLogPG::truncate_update_size_and_usage(
+  object_stat_sum_t& delta_stats,
+  object_info_t& oi,
+  uint64_t truncate_size)
+{
+  if (oi.size == truncate_size) {
+    // no change
+    return;
+  }
+  if (oi.has_extents()) {
+    int64_t old_bytes = oi.extents.size();
+    if (truncate_size > oi.size) {
+      // trunc up
+      interval_set<uint64_t> to_add;
+      to_add.insert(oi.size, truncate_size - oi.size);
+      oi.extents.union_of(to_add);
```

Comment:
should a simple oi.extents.insert(oi.size, truncate_size - oi.size) be sufficient?  no need for to_add and union_of